File: repository/src/main/java/by/rekuts/travelagency/repository/impl/UserRepositoryImpl.java
package by.rekuts.travelagency.repository.impl;

import by.rekuts.travelagency.domain.Tour;
import by.rekuts.travelagency.domain.User;
import by.rekuts.travelagency.domain.aspects.LogReturn;
import by.rekuts.travelagency.repository.Specification;
import by.rekuts.travelagency.repository.UserRepository;
import by.rekuts.travelagency.repository.UserSpecification;
import org.springframework.stereotype.Repository;

import javax.persistence.EntityManager;
import javax.persistence.PersistenceContext;
import javax.persistence.criteria.Predicate;
import java.util.List;

@Repository
public class UserRepositoryImpl implements UserRepository {
    @PersistenceContext
    private EntityManager entityManager;

    @Override
    public void insert(User user) {
        if(user.getRole() == null) {
            user.setRole(User.UserRole.ROLE_MEMBER);
        }
        entityManager.persist(user);
    }

    @Override
    public void update(User user) {
        var dbUser = entityManager.find(User.class, user.getId());
        dbUser.setLogin(user.getLogin());
        dbUser.setPassword(user.getPassword());
        dbUser.setRole(user.getRole());
        entityManager.merge(dbUser);
    }

    @Override
    public void delete(int id) {
        entityManager.remove(entityManager.find(User.class, id));
    }

    @LogReturn
    @Override
    public List<User> getList(Specification specification) {
        var userSpecification = (UserSpecification) specification;
        var builder = entityManager.getCriteriaBuilder();
        var criteriaQuery = builder.createQuery(User.class);
        var root = criteriaQuery.from(User.class);
        List<Predicate> predicates;
        predicates = userSpecification.getPredicates(root, builder);

        if (!predicates.isEmpty()) {
            criteriaQuery.where(
                    predicates.toArray(new Predicate[]{})
            );
        }
        return entityManager.createQuery(criteriaQuery).getResultList();
    }

    public void addTourToFavorites(User user, Tour tour) {
        var tours = user.getTours();
        if (!tours.contains(tour)) {
            tours.add(tour);
        }
        user.setTours(tours);
        entityManager.merge(user);
    }

    public void removeTourFromFavorites(User user, Tour tour) {
        var tours = user.getTours();
        tours.remove(tour);
        user.setTours(tours);
        entityManager.merge(user);
    }
}

File: repository/src/main/scala/by/rekuts/travelagency/repository/impl/UserRepositoryImpl.scala
package by.rekuts.travelagency.repository.impl

import java.util

import by.rekuts.travelagency.domain.{Tour, User, UserRole}
import by.rekuts.travelagency.repository.{Specification, UserRepository, UserSpecification}
import javax.persistence.criteria.{CriteriaBuilder, CriteriaQuery, Predicate, Root}
import org.springframework.stereotype.Repository

@Repository
class UserRepositoryImpl extends UserRepository {

  override def insert(user: User): Unit = {
    if(user.role == null) {
      user.role_$eq(UserRole.ROLE_MEMBER)
    }
    entityManager.persist(user)
  }

  override def update(user: User): Unit = {
    val dbUser = entityManager.find(classOf[User], user.id)
    dbUser.login_$eq(user.login)
    dbUser.password_$eq(user.password)
    dbUser.role_$eq(user.role)
    entityManager.merge(dbUser)
  }

  override def delete(id: Int): Unit = entityManager.remove(entityManager.find(classOf[User], id))

  override def getList(specification: Specification[_]): util.List[User] = {
    val reviewSpecification: UserSpecification = specification.asInstanceOf[UserSpecification]
    val builder: CriteriaBuilder = entityManager.getCriteriaBuilder
    val criteriaQuery: CriteriaQuery[User] = builder.createQuery(classOf[User])
    val root: Root[User] = criteriaQuery.from(classOf[User])
    val predicates = reviewSpecification.getPredicates(root, builder)
    if (!predicates.isEmpty) {
      criteriaQuery.where(predicates.toArray[Predicate](new Array[Predicate](predicates.size())): _*)
    }
    entityManager.createQuery(criteriaQuery).getResultList
  }

  override def addTourToFavorites(user: User, tour: Tour): Unit = {
    val tours = user.tours
    if (!tours.contains(tour)) tours.add(tour)
    user.tours_$eq(tours)
    entityManager.merge(user)
  }

  override def removeTourFromFavorites(user: User, tour: Tour): Unit = {
    val tours = user.tours
    tours.remove(tour)
    user.tours_$eq(tours)
    entityManager.merge(user)
  }
}
